### ISSUE TITLE
Widget size issues on different Android launchers

### ISSUE THREAD

**Issue by sAleksovski** [sAleksovski/react-native-android-widget#34]
We cannot render React Native views directly to the widget.
What this library does is render the React Native views to an image, and then show that image in the widget.
For it to look good, we need to know the **exact** size of the widget, so we can create an image that will fit the widget correctly.

On some Android launchers, the reported and actual size of the widget are not always the same.

![Actual size](https://user-images.githubusercontent.com/7473800/244463696-2c017d1a-34b8-4a20-9f2e-22b4a3d9a739.png)
Actual size (365dp x 318dp)

![Reported size](https://user-images.githubusercontent.com/7473800/244463758-534f54a0-f3d2-4a0f-aaaa-0f2c5fc259b8.png)
Reported size (365dp x 354dp)

I haven't found a reliable way to get the **exact** size, the current behavior is to crop the widget if the reported size is smaller.

**Comment by OPavliuk-delhaize** [comment]
We have a problem the size of the widget does not come at all, it returns width:0 and height: 0. Of all the tested phones, this is observed only on **Samsungs**. 

Debugging showed that the value 0 is returned by the method

```
public static int getWidgetSizeInDp(Context context, int widgetId, String key) {
 return AppWidgetManager.getInstance(context).getAppWidgetOptions(widgetId).getInt(key, 0);
 }
```
in **java/com/reactnativeandroidwidget/RNWidgetUtil.java**

Do you have ideas about what could be the problem? 
Looks like some sort of asynchronous code-handling error.

We've minimized the impact of this bug by setting default sizes for width and height, but even once valid non-zero values ​​are received, they are stored in the app and used in the widget.

**Comment by sAleksovski** [comment]
Does it received 0 only when first added? 

I don't know any other way to get the widget size.
You can try to request another widget update if you receive 0 sizes.

Btw I have a Samsung phone and I receive correct sizes.

**Comment by OPavliuk-delhaize** [comment]
Yes, 0 comes precisely for adding the widget. But this does not always happen.

It's an interesting idea to update the widget again if the value is 0, I'll try.

Maybe you're lucky. We have this problem consistently on Samsung S24+ and Samsung Galaxy A13 Android 14

**Comment by colonelpanic8** [comment]
I can reproduce this with `react-native-android-widget` 0.21.0 on a Samsung foldable (`SM-F976U1`, Android API 37) using Lawnchair 15 Beta 3 while the device is unfolded.

Lawnchair provides these options for the widget instance:

```text
appWidgetMinWidth=290
appWidgetMaxWidth=337
appWidgetMinHeight=286
appWidgetMaxHeight=303
appWidgetSizes=[
  337.66666x303.0,
  337.66666x303.0,
  290.66666x286.66666,
  290.66666x286.66666
]
```

The current portrait logic in `RNWidgetUtil` combines `OPTION_APPWIDGET_MIN_WIDTH` with `OPTION_APPWIDGET_MAX_HEIGHT`, so it renders a `290x303dp` bitmap. That pair is not one of the sizes supplied by the launcher. The host is displaying the `337.67x303dp` size, leaving an approximately 47dp blank strip on the right. A root view using `width: "match_parent"` therefore still does not fill the resize frame.

This is distinct from a launcher returning an inaccurate size: Lawnchair is providing the Android 12 `OPTION_APPWIDGET_SIZES` list for its supported foldable profiles, while the library is synthesizing a size by combining independent extrema. The same `RNWidgetUtil` implementation is still present in 0.22.0.

As a narrow compatibility workaround, I changed the selection logic to:

1. Preserve the current min/max-derived size if that exact pair appears in `OPTION_APPWIDGET_SIZES`.
2. Preserve the current behavior on pre-Android 12 devices or when the explicit list is absent.
3. If the derived pair is not an advertised size, use an explicit launcher-provided size instead.

That fixes this instance without changing behavior for the other launchers tested, which either provide a single explicit size or include the legacy-derived pair in their list. A complete upstream solution may need to render exact `RemoteViews` variants for all entries in `OPTION_APPWIDGET_SIZES`, as recommended by the Android 12 widget sizing API.

Android documentation: https://developer.android.com/develop/ui/views/appwidgets/layouts

**Comment by outaTiME** [comment]
**Environment**: react-native-android-widget 0.22.1, RN 0.86.2, Expo SDK 57. Provider: `minWidth/minHeight=110dp`, `targetCellWidth/Height=2`, `resizeMode="none"`, `widgetFeatures="reconfigurable|configuration_optional"`.

I hit this on two physical devices and measured it end to end, in case the numbers help.

Method: I pulled the PNG the library writes to the app's `files/widget_images/` folder with `adb run-as` from a debug build, which is the bitmap it generated, and got the real on-screen size of the widget from `adb shell uiautomator dump`, which gives the bounds of the `AppWidgetHostView` and of the `ImageView` that displays that bitmap.

| | reported (options) | generated bitmap | actual `ImageView` | what I see |
|---|---|---|---|---|
| Galaxy S9, Android 10 | 166x289 dp | 166x289 dp | 144x241 dp | widget renders at ~78% of the bitmap, looks small in its cell |
| Moto Edge 30 Ultra, Android 15 | 208x277 dp | 208x277 dp | ~192 dp wide | cropped 16dp on the right, rounded corners lost on that side |

In both cases the generated bitmap matches the reported size exactly.

One difference between the two devices when adding a widget: on the S9 I get `WIDGET_ADDED` with 174x297 and then `WIDGET_RESIZED` with 166x289. On the Moto the configuration screen never opens and only `WIDGET_ADDED` fires, with 208x277 — no second event with a corrected size.

Also, since API 31 there's `AppWidgetManager.OPTION_APPWIDGET_SIZES`, a `List<SizeF>` with the sizes in dp the widget can take. Google's docs say about the `OPTION_APPWIDGET_MIN_WIDTH` / `MAX_HEIGHT` approach that *"that logic doesn't work in all situations"*. I couldn't test whether it reports something different on these devices.

**Comment by CatLover01** [comment]
> Yes, 0 comes precisely for adding the widget. But this does not always happen.
> 
> It's an interesting idea to update the widget again if the value is 0, I'll try.
> 
> Maybe you're lucky. We have this problem consistently on Samsung S24+ and Samsung Galaxy A13 Android 14

Interesting, we are getting 0 on app update (e.g. widget doesn't render anything visually), reproduced on Samsung A32 but S20 doesn't do it. It's probably slower devices of which this is happening on.
